File: baekjoon/python/silver1/q21608_sharks_elementary_school.py
## https://www.acmicpc.net/problem/21608
def solution():
    n = int(input())
    board = [[0] * n for _ in range(n)]
    move = [(-1, 0), (0, 1), (1, 0), (0, -1)]
    like_data = [[] for _ in range(n**2+1)]
    for _ in range(n**2):
        next_x = -1
        next_y = -1
        data = list(map(int, input().split()))
        s = data[0]
        like_data[s] = data
        like = data[1:]
        candidate = []
        like_number = 0
        for i in range(n):
            for j in range(n):
                if board[i][j] == 0:
                    cnt = 0
                    for k in range(4):
                        nx = i+move[k][0]
                        ny = j+move[k][1]
                        if 0<= nx < n and 0 <= ny < n:
                            if board[nx][ny] in like:
                                cnt += 1
                    if cnt == like_number :
                        candidate.append((i, j))
                    elif cnt > like_number :
                        candidate = [(i, j)]
                        like_number = cnt
        if len(candidate) > 1 : 
            new_candidates = []
            adjacent_emty = 0
            for can in candidate:
                i = can[0]
                j = can[1]
                cnt = 0
                for k in range(4):
                        nx = i+move[k][0]
                        ny = j+move[k][1]
                        if 0<= nx < n and 0 <= ny < n:
                            if board[nx][ny] == 0 :
                                cnt += 1
                if cnt == adjacent_emty :
                    new_candidates.append((i, j))
                elif cnt > adjacent_emty:
                    new_candidates = [(i, j)]
                    adjacent_emty = cnt
            candidate = new_candidates
            if len(candidate) > 1:
                candidate.sort()
        next_x = candidate[0][0]
        next_y = candidate[0][1]
        board[next_x][next_y] = s

        statisfaction = 0
        for i in range(n):
            for j in range(n):
                x = board[i][j]
                cnt = 0
                for k in range(4):
                    nx = i+move[k][0]
                    ny = j+move[k][1]
                    if 0<= nx < n and 0 <= ny < n:
                        if board[nx][ny] in like_data[x] :
                            cnt += 1
                if cnt != 0:
                    statisfaction += 10 ** (cnt-1)
        
    return statisfaction    

print(solution())

## 파이썬 랭커 id: jh05013 님의 풀이 
def other():
    n = int(input())
    grid = [[None]*n for i in range(n)]
    likes = [[] for i in range(n**2+1)]
    for ITER in range(n**2):
        no, *like = map(int,input().split())
        likes[no] = like
        ans = (9999, 9999, -1, -1) # -like, -emp, i, j
        for i in range(n):
            for j in range(n):
                if grid[i][j]: continue
                l0 = emp0 = 0
                for ni,nj in (i-1,j),(i+1,j),(i,j-1),(i,j+1):
                    if not (0<=ni<n and 0<=nj<n): continue
                    if grid[ni][nj] == None: emp0+= 1
                    elif grid[ni][nj] in like: l0+= 1
                ans = min(ans, (-l0, -emp0, i, j))
        _, __, i, j = ans
        grid[i][j] = no

    ans = 0
    for i in range(n):
        for j in range(n):
            like = likes[grid[i][j]]
            l0 = 0
            for ni,nj in (i-1,j),(i+1,j),(i,j-1),(i,j+1):
                if 0<=ni<n and 0<=nj<n and grid[ni][nj] in like: l0+= 1
            ans+= [0, 1, 10, 100, 1000][l0]
    print(ans)

print(other())

File: baekjoon/java/src/silver1/Q21608_SharksElementarySchool.java
package silver1;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

//https://www.acmicpc.net/problem/21608
public class Q21608_SharksElementarySchool {
    // 파이썬 랭커 jh05013 풀이 응용
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[][] board = new int[n][n];
        List<List<Integer>> likes = new ArrayList<>((int) (Math.pow(n, 2) + 1));
        for (int i = 0; i <= Math.pow(n, 2); i++) {
            likes.add(new ArrayList<>());
        }
        for (int k = 0; k < Math.pow(n, 2); k++) {
            List<Integer> data = Arrays.stream(br.readLine().split(" ")).map(Integer::parseInt).collect(Collectors.toList());
            int num = data.get(0);
            data = data.subList(1, data.size());
            likes.set(num, data);
            Integer[] ans = {-1, -1, -1, -1};
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (board[i][j] != 0) {
                        continue;
                    }
                    int like = 0;
                    int empty = 0;
                    for (int[] move : new int[][]{{i - 1, j}, {i, j + 1}, {i + 1, j}, {i, j - 1}}) {
                        int nx = move[0];
                        int ny = move[1];
                        if (0 <= nx && nx < n && 0 <= ny && ny < n) {
                            if (board[nx][ny] == 0) {
                                empty++;
                            }
                            if (data.contains(board[nx][ny])) {
                                like++;
                            }
                        }
                    }
                    Integer[] now = new Integer[]{like, empty, i, j};
                    List<Integer[]> temp = new ArrayList<>();
                    temp.add(ans);
                    temp.add(now);
                    temp.sort((a, b) -> {
                        if (a[0] == b[0]) {
                            if (a[1] == b[1]) {
                                if (a[2] == b[2]) {
                                    return a[3] - b[3];
                                }
                                return a[2] - b[2];
                            }
                            return b[1] - a[1];
                        }
                        return b[0] - a[0];
                    });
                    ans = temp.get(0);
                }
            }
            board[ans[2]][ans[3]] = num;
        }

        int answer = 0;
        int[] point = new int[]{0, 1, 10, 100, 1000};
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                List<Integer> like = likes.get(board[i][j]);
                int lp = 0;
                for (int[] move : new int[][]{{i - 1, j}, {i, j + 1}, {i + 1, j}, {i, j - 1}}) {
                    int nx = move[0];
                    int ny = move[1];
                    if (0 <= nx && nx < n && 0 <= ny && ny < n) {
                        if (like.contains(board[nx][ny])) {
                            lp++;
                        }
                    }
                }
                answer += point[lp];
            }
        }
        System.out.println(answer);
    }
}
